packages/shylu/shylu_dd/frosch/src/Adapters/FROSch_TpetraPreconditioner_def.hpp

```diff
@@ -0,0 +1,159 @@
+
+    template <class SC,class LO,class GO,class NO>
+    TpetraPreconditioner<SC,LO,GO,NO>::TpetraPreconditioner()
+    {
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    TpetraPreconditioner<SC,LO,GO,NO>::TpetraPreconditioner(Teuchos::RCP<SchwarzPreconditioner<SC,LO,GO,NO> > preconditioner) :
+    preconditioner_(preconditioner)
+    {
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    TpetraPreconditioner<SC,LO,GO,NO>::~TpetraPreconditioner()
+    {
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    int TpetraPreconditioner<SC,LO,GO,NO>::
+    initialize(bool useDefaultParameters)
+    {
+      return preconditioner_->initialize(useDefaultParameters);
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    int TpetraPreconditioner<SC,LO,GO,NO>::
+    compute()
+    {
+      return preconditioner_->compute();
+    }
+
+    // Y = alpha * A^mode * X + beta * Y
+    template <class SC,class LO,class GO,class NO>
+    void TpetraPreconditioner<SC,LO,GO,NO>::
+    apply(const Tpetra::MultiVector<SC,LO,GO,NO> &X,
+                Tpetra::MultiVector<SC,LO,GO,NO> &Y,
+          ETransp mode,
+          SC alpha,
+          SC beta) const
+    {
+      using XMultiVector = Xpetra::TpetraMultiVector<SC,LO,GO,NO>;
+      auto xTmp = const_cast<Tpetra::MultiVector<SC,LO,GO,NO>&>(X);
+      auto xRef = Xpetra::rcpFromRef(xTmp);
+      const XMultiVector xX(xRef);
+            XMultiVector xY(Xpetra::rcpFromRef(Y));
+      preconditioner_->apply(xX, xY, mode, alpha, beta);
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    Teuchos::RCP<const Tpetra::Map<LO,GO,NO> > TpetraPreconditioner<SC,LO,GO,NO>::
+    getDomainMap() const
+    {
+      return Xpetra::toTpetraNonZero(preconditioner_->getDomainMap());
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    Teuchos::RCP<const Tpetra::Map<LO,GO,NO> > TpetraPreconditioner<SC,LO,GO,NO>::
+    getRangeMap() const
+    {
+      return Xpetra::toTpetraNonZero(preconditioner_->getRangeMap());
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    void TpetraPreconditioner<SC,LO,GO,NO>::
+    describe(FancyOStream &out,
+             const EVerbosityLevel verbLevel) const
+    {
+      return preconditioner_->describe(out, verbLevel);
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    string TpetraPreconditioner<SC,LO,GO,NO>::
+    description() const
+    {
+      return preconditioner_->description();
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    bool TpetraPreconditioner<SC,LO,GO,NO>::isInitialized() const
+    {
+      return preconditioner_->isInitialized();
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    bool TpetraPreconditioner<SC,LO,GO,NO>::isComputed() const
+    {
+      return preconditioner_->isComputed();
+    }
+
+    template <class SC,class LO,class GO,class NO>
+    void TpetraPreconditioner<SC,LO,GO,NO>::residual(const TMultiVector & X,
```

Comment:
This is correct, but what is this needed for? It might make sense to throw instead.